File: app/models/ability.rb
class Ability < ApplicationRecord
  FEATURES = %w(scan_id_card liveness_test)
  DEFAULT_USE_COUNTS = {
    "scan_id_card" => 3,
    "liveness_test" => 0
  }
  CHECK_ABILITY = false

  serialize :use_counts, Hash

  def abilities
    FEATURES.reduce({}) do |ret, feature|
      ret[feature] = has_ability?(feature)
      ret
    end
  end

  def use_count feature
    use_counts[feature] || DEFAULT_USE_COUNTS[feature] || 0
  end

  def has_ability? feature
    return true unless CHECK_ABILITY
    use_count(feature) > 0
  end

  def decrease_use_count! feature
    return true unless CHECK_ABILITY
    use_counts[feature] = use_count(feature) - 1
    save!
  end
end

File: web/models/ability.ex
defmodule Scanner.Ability do
  use Scanner.Web, :model

  schema "abilities" do
    field :unique_id, :string
    field :use_counts, :map, default: %{}

    timestamps()
  end

  @abilities ~w(scanIDCard livenessTest)

  def abilities ability do
    @abilities
    |> Enum.map(&({&1, has_ability?(ability, &1)}))
    |> Enum.into(%{})
  end

  def use_count ability, feature do
    ability.use_counts[feature] || 0
  end

  def decrease_use_count ability, feature do
    use_counts =
      ability.use_counts
      |> Map.put(feature, use_count(ability, feature) - 1)
    changeset(ability, %{use_counts: use_counts})
  end

  def has_ability? ability, feature do
    use_count(ability, feature) > 0
  end

  @doc """
  Builds a changeset based on the `struct` and `params`.
  """
  def changeset struct, params \\ %{} do
    struct
    |> cast(params, [:unique_id, :use_counts])
    |> validate_required([:unique_id, :use_counts])
  end
end
